### PR TITLE
RFC: mps: fix batch norm for non-contiguous channels-last inputs

### PR DESCRIPTION
**Pull request by roei-shlezinger**
Fix for https://github.com/pytorch/pytorch/issues/156792.

The bug occurs when using MPS BatchNorm with inputs that are both
`channels_last` *and* non-contiguous views (e.g. after `permute`).
The MPSGraph was executed in an implicit NCHW layout, while the axes and
placeholder shapes were derived from the original NHWC tensor, leading to
incorrect reduction axes and wrong gradients in backward.

This patch keeps the internal graph execution in a consistent
contiguous/NCHW layout, and handles channels-last only at the boundary
via reshapes. It also fixes `outputNeedsCopy` to avoid unnecessary
allocations/copies for channels_last-contiguous outputs.

A test covering this exact failure case has been added — the test now passes.

### CODE REVIEW

**Comment by roei-shlezinger** [comment]
Thanks for the pointer to #162776 — I skimmed it now and will read it more thoroughly soon. 

Just for context: I'm on macOS 26.1 (Apple Silicon, current `main`), and the BatchNorm issue still reproduces for me. The new regression test here fails without this patch and passes with it, so it seems there is still a correctness gap even with the newer strided-tensor support.

If it's reasonable to land this as-is and handle any macOS-15+ improvements in a follow-up, I can do that.  
If you'd prefer changes in this PR instead, I'm happy to update it.  
Please let me know which direction you think is best.

**Comment by roei-shlezinger** [comment]
Applied your suggestion: dropped macOS 14 support and simplified the macOS 15 path (net reduction overall); would appreciate another look. Thanks!

**Comment by benpresles** [comment]
I am running macOS 15.7.3 with PyTorch 2.9.1 (Python 3.12.12), and I am experiencing the same issue as you. Will your fix be implemented in the next release of PyTorch?

**Comment by kurtamohler** [comment]
I don't think we want to fully drop macOS 14 support yet. Instead, we should probably keep the current behavior for macOS 14 and use the MPSGraph strided tensors for macOS 15

**Comment by benpresles** [comment]
I have just tried with the new version of PyTorch (2.10.0) and I am still experiencing the same issue. Do you know when a fix will be available?

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 1]
```diff
@@ -87,288 +88,544 @@ static void get_shapes(MPSShape* input_shape_readonly,
   TORCH_CHECK_NOT_IMPLEMENTED(self.scalar_type() != kLong, "Long batch norm is not supported with MPS");
   TORCH_CHECK_NOT_IMPLEMENTED(!c10::isComplexType(self.scalar_type()),
                               "Batch norm for complex is not supported for MPS");
-  using namespace at::native::mps;
-  struct CachedGraph : public MPSCachedGraph {
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/Normalization.mm:121]: Sorry to ask for another large change, but the diff is pretty difficult to read as a human, because a lot of the old code is being modified to add extra indentation, so there are lots of deleted lines interleaved with the added lines.

~IMO, it would be better to rename the old `batch_norm_mps_out` to `batch_norm_mps_out_macos_14` (or some similar name), then add the new macos 15 impl as a new function `batch_norm_mps_out_macos_15`, and then make `batch_norm_mps_out` conditionally call either the macos 14 or 15 function. That way the diff will barely touch the old code and will mostly just be one block of added lines. Then also in the future if we decide to remove the old impl, we won't have to touch much of the new code either, we'll just delete the function for the old impl. Does that make sense?~

**Review by kurtamohler — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mps/operations/Normalization.mm (hunk 1 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/operations/Normalization.mm:121]: Actually, we probably shouldn't do what I said above. Looking at the new and old impls side to side, it looks like the vast majority of the code is the same, and there are only differences in a few places. How about instead of having two completely separate impls, we can have one impl but in any spot where we need to do something different for either version we have a conditional.

**Comment by benpresles** [comment]
Hi @roei-shlezinger,
By any chance, have you had time to address the review comments on this PR?
